Add sample application and README file

JatinSanghvi opened the pull request:
*Checking the PR branch in forked repo: https://github.com/JatinSanghvi/keda-external-scaler-azure-cosmos-db (including the README file) might provide a better sense of the changes that were made*

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)

### Fixes

- Addresses issue https://github.com/kedacore/keda-external-scaler-azure-cosmos-db/issues/2
- Addresses issue https://github.com/kedacore/keda-external-scaler-azure-cosmos-db/issues/3
- Addresses issue https://github.com/kedacore/keda-external-scaler-azure-cosmos-db/issues/9 *(need review to confirm)*

- **Scales targets based on count of Cosmos DB container's partition-ranges with non-zero pending items**. Existing implementation scales according to the estimate of items pending to be processed across all partitions, which is incorrect. Cosmos DB architecture limits the degree of parallelism of "processing new changes to the container" to the "number of logical partition ranges" ([More information](https://docs.microsoft.com/en-us/azure/cosmos-db/change-feed#working-with-change-feed)).  The logical partition ranges is dependent on the physical partitions, which is turn depends on combination of container data storage size and container throughput ([More information](https://docs.microsoft.com/en-us/azure/cosmos-db/partitioning-overview#physical-partitions)). For small containers with low throughput provisioned, we expect the scale targets to not scale beyond 1.

- **Fixes generation of metric names**. Existing implementation generates the metric name in method `GetMetricSpec` but uses metric name from request metadata in method `GetMetrics`. If metric name is not specified in `ScaledObject` definition, returning metric name in `GetMetricSpec` won't make the metric-name part of scaler metadata when `GetMetrics` is called. On other hand, if the metric name is supplied by user in `ScaledObject` definition, `GetMetricSpec` would override the name, and KEDA won't be able to get the metrics for the generated metric-name when calling `GetMetrics`. New implementation provides consistent behavior across methods. All methods will reuse the metric-name if provided by user, else generate one based on other 'required' properties in `ScaledObject`.

- **Bases metric name generation based on lease container properties**. The current metric-name generation logic will create same metric name if two different lease containers are created to monitor the same container.

- **Uses standard names as in Cosmos DB documentation for trigger properties**. `Connection` instead `ConnectionString` (matches with other KEDA scalers), `ContainerId` instead of `CollectionName` and `ProcessorName` instead of `LeaseCollectionPrefix`.

### Other changes

- Adds sample producer and consumer projects with steps in README file to test scaling of target application between 0, 1 and 2 replicas.
- Refactors code and updates unit tests.

JatinSanghvi commented on .gitignore:
```diff
@@ -4,6 +4,7 @@
 ## Get latest from https://github.com/github/gitignore/blob/master/VisualStudio.gitignore
 
 # User-specific files
+*.rsuser
```
Updated file to one generated by `dotnet new gitignore`.

JatinSanghvi commented on .github/workflows/main-build.yml:
```diff
@@ -0,0 +1,79 @@
+          tags: |
+            type=ref,event=pr
+            type=ref,event=tag
+            type=raw,value=${{ env.VERSION_TAG }}
```
A new tag, along with creating versioned tag for the GHCR image, will also override the `experimental` tag. I will look into this.

JatinSanghvi replied:
Resolved to some extent. There will be no image with tag `latest` on new repo tag. It should be okay to live with overriding behavior of `experimental` tag for now, since in most cases, the latest commit on `main` branch will be tagged.

JatinSanghvi commented:
Tagging @pragnagopa, @TsuyoshiUshio, @divyagandhisethi, @ealsur, @lpapudippu

ealsur commented on Keda.CosmosDbScaler.Demo.OrderProcessor/Worker.cs:
```diff
@@ -0,0 +1,79 @@
+            // Change feed processor instance name should be unique for each container application.
+            string instanceName = $"Instance-{Dns.GetHostName()}";
+
+            _processor = new CosmosClient(_cosmosDbConfig.Connection)
```
if the Connection of the lease is the same (LeaseConnection == Connection) you don't need 2 clients

JatinSanghvi replied:
Since this one is a demo application, I thought to not add an extra if-check for conciseness and readability. But, since the order-processor application might stay running for long time and would keep referencing the `CosmosClient` instances during the lifetime, I will get it done. BTW, the scaler application has `CosmosDbFactory` class that ensures that the `CosmosClient` instance is re-used.

JatinSanghvi replied:
Skipping this one since it would complicate the implementation. The code run in a sample application; it won't be packaged, deployed or be part of customer's production services.

ealsur commented on Keda.CosmosDbScaler.Demo.OrderProcessor/Worker.cs:
```diff
@@ -0,0 +1,79 @@
+            string instanceName = $"Instance-{Dns.GetHostName()}";
+
+            _processor = new CosmosClient(_cosmosDbConfig.Connection)
+                .GetDatabase(_cosmosDbConfig.DatabaseId)
```
No need to call GetDatabase().GetContainer(). You can call client.GetContainer() directly

ealsur commented on Keda.CosmosDbScaler.Demo.OrderProcessor/Worker.cs:
```diff
@@ -0,0 +1,79 @@
+        private async Task ProcessOrdersAsync(IReadOnlyCollection<Order> orders, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation(orders.Count + " orders received");
+            var tasks = new List<Task>();
```
what's the task list for?

JatinSanghvi replied:
Thanks for pointing out. Must be residual code. I will delete the variable.

ealsur commented on Keda.CosmosDbScaler/Services/CosmosDbMetricProvider.cs:
```diff
@@ -0,0 +1,75 @@
+                    while (iterator.HasMoreResults)
+                    {
+                        FeedResponse<ChangeFeedProcessorState> states = await iterator.ReadNextAsync();
+                        partitionCount += states.Where(state => state.EstimatedLag > 0).Count();
```
Why only counting the number of states that have lag > 0? Each state represents 1 lease, so the count of the partitions is = to the number of states regardless of current lag

JatinSanghvi replied:
@ealsur, though the variable name is `partitionCount`, it is the count of partitions that have one or more items pending to be processed. The value returned by the method governs how many hosts will be spawned as in [this diagram](https://docs.microsoft.com/en-us/azure/cosmos-db/sql/change-feed-processor#components-of-the-change-feed-processor). For example, if number of partitions is 2, but there are no items pending for consumption, the method should return 0.

I created a test-application to confirm that number of states would be non-zero even if there are no pending items. That is not the number that the method intends to return. Feel free to reopen the conversation if this requires more discussion.

ealsur replied:
@JatinSanghvi Makes sense if that is how the value is used. If the value was used as "This is the max number of instances we can scale up to" that's when I saw an issue with the filtering, because really the max number of instances is the number of states/partitions, regardless of pending changes.

JatinSanghvi commented:
@ahmelsayed, @tomkerkhove - Can you please review the PR?

pragnagopa commented on .vscode/launch.json:
```diff
@@ -0,0 +1,35 @@
+{
+    "version": "0.2.0",
+    "configurations": [
```
add files in .vscode to ignore list as well

JatinSanghvi replied:
We are only including essential `.vscode` files. The list of essential files is there in `.gitignore`.

pragnagopa commented on Keda.CosmosDbScaler/Protos/externalscaler.proto:
```diff
@@ -1,51 +1,45 @@
-	map<string, string> metadata = 2;
+syntax = "proto3";
+
+option csharp_namespace = "Keda.CosmosDbScaler";
```
any strong reason to update the naming convention from `CosmosDb.Scaler` to `CosmosDBScaler` ?

JatinSanghvi replied:
The namespace-tree was shortened to keep the count of namespace-branches in the test of demo projects smaller. Since the namespace gets reused for project-file name and assembly name, this helps reducing extra dots in the filenames. If we need to use three-node namespace, I would prefer `Keda.Scalers.CosmosDb` instead of `Keda.CosmosDb.Scaler` as we won't have any other namespace with `Keda.CosmosDb.` as prefix, but we can have `Keda.Scalers.<event-source>` as namespace in a different `kedacore` repo.

JatinSanghvi commented:
@ahmelsayed, @tomkerkhove - Gentle reminder for PR review.

tomkerkhove commented:
I just got back from holiday and will review asap

tomkerkhove commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,79 @@
+name: Build
```
We have 2 kinds of artifacts here:
1. Container for the external scaler that needs SemVer
2. Containers to help as part of samples

Personally I would split them into two workflows given the sample containers can be pushed every commit and don't have to follow the SemVer of the external scaler itself

JatinSanghvi replied:
The GitHub action builds both scaler and sample images but only pushes the scaler image. Since there is no known issue with having a single pipeline, I am leaving it so until we find a limitation with it in future.

I will be working on your README and code-structuring comments. I am leaning more towards naming the namespace as `Keda.Scalers.CosmosDb` so that we can have namespaces in other scaler repos as `Keda.Scalers.AnotherEventType` in future. Let me know your preference.

tomkerkhove replied:
I don't have must of preference on the naming but if we have other scalers, I'd argue that they have to end up in a new repo since they will have their own versioning scheme which would be confusing imo

JatinSanghvi replied:
@tomkerkhove - I still am not clear why we should document that this is purely for change feed. I am not aware of any other way Cosmos DB supports polling for new changes to the Container. All listener applications are supposed to have change feeds setup. Let me know if you find a link or reference that proves this to be incorrect.

tomkerkhove replied:
Because there is more to autoscaling Cosmos DB apps than just changefeed :)

People can autoscale based on SQL queries, for example. You are not required to use changefeed, it is just one way of using Cosmos DB.

JatinSanghvi replied:
@tomkerkhove - I think there is misunderstanding on your end. Change feeds are not meant to be auto-scaled. They are mechanism provided by Cosmos DB to track changes to the Cosmos containers.

JatinSanghvi replied:
OK I assumed the apps being auto-scaled are always trigger-based. That they would only be interested in new updates to the data, instead of looking for an ability to be auto-scaled based on output of a query that runs on the entire data (not just the changes).

tomkerkhove replied:
Yes, they are just different use-cases (and both are fine) but we just have to be explicit and set the right expectations for end-users.

JatinSanghvi replied:
Addressed comments.

tomkerkhove commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,79 @@
+
+      # Using metadata ensures all lower-case tag-name even if the GitHub username has upper-case letters.
+      - name: Build (and push on commit) Docker image for scaler
+        uses: docker/build-push-action@v2
```
```suggestion
        if: github.event_name == 'push'
        uses: docker/build-push-action@v2
```

JatinSanghvi replied:
Hi @tomkerkhove, line 79 in `build.yml` applies the same filter as the one you suggested.

tomkerkhove replied:
That adds the condition to the log in but not on pushing so it will fail because it will try to run it

JatinSanghvi replied:
Line 55 has the condition for login. Line 79 takes care of conditional push.

tomkerkhove commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,79 @@
+
+      # Using metadata ensures all lower-case tag-name even if the GitHub username has upper-case letters.
+      - name: Build (and push on commit) Docker image for scaler
+        uses: docker/build-push-action@v2
```
You should split this step into `build` and `push` because you will always want to build the image, but only push for main commits (experimental) / releases (SemVer)

JatinSanghvi replied:
This is already taken care of. The condition at line 79 ensures the push is done only for commits and new release-tags.

tomkerkhove reviewed (changes requested):
I did a first skip but have a few high-level remarks:
- The README does not provide documentation on what the scaler itself does and how it helps the customer. You should provide guidance on what it is, how to deploy and how to configure a ScaledObject
- The scaler project was renamed, what was the reasoning for this as the previous naming convention (`CosmosDb.Scaler`) was a lot better imo
- The repo needs a bit more structure, can you please introduce the following folders please?
  - `src/` for the external scaler code
  - `samples/` for the sample code and README to provide guidance how to run the sample
- We should explicitly document that this is purely for Changefeed and not the rest

JatinSanghvi commented:
@tomkerkhove - Please review the updated PR when you have time.

- Added main README file with scalar overview, deployment instructions and trigger metadata details. Moved earlier README for sample app to sub-directory.
- Introduced separate GitHub actions for different triggers to avoid confusion and ensure the built package versions do not overlap.
- Updated namespace names from `Keda.CosmoDbScaler` to `Keda.CosmoDb.Scaler`.
- Updated folder structure. The sample app is moved to `/src/Scaler.Demo` as that made more sense to me and helped retaining all project references in the solution file.

tomkerkhove commented on src/Scaler.Demo/Shared/CosmosDbConfig.cs:
```diff
@@ -0,0 +1,22 @@
+        public string ContainerId { get; set; }
+        public int ContainerThroughput { get; set; }
+
+        public string LeaseConnection { get; set; }
+        public string LeaseDatabaseId { get; set; }
+        public string LeaseContainerId { get; set; }
```
Wouldn't it be better if we introduce a `Lease` object that has these fields?
```suggestion
        public string LeaseConnection { get; set; }
        public string LeaseDatabaseId { get; set; }
        public string LeaseContainerId { get; set; }
```

tomkerkhove commented on src/Scaler.Demo/Shared/CosmosDbConfig.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public string Connection { get; set; }
+        public string DatabaseId { get; set; }
+        public string ContainerId { get; set; }
+        public int ContainerThroughput { get; set; }
```
Wouldn't it be better if we introduce a `Container` object that has these fields?
```suggestion
        public string ContainerId { get; set; }
        public int ContainerThroughput { get; set; }
```

tomkerkhove commented on src/Scaler/Services/ScalerMetadata.cs:
```diff
@@ -0,0 +1,55 @@
+        public string Connection { get; set; }
+        public string DatabaseId { get; set; }
+        public string ContainerId { get; set; }
+        public string LeaseConnection { get; set; }
+        public string LeaseDatabaseId { get; set; }
+        public string LeaseContainerId { get; set; }
```
Wouldn't it be better to introduce a `Lease` object that has these as properties?
```suggestion
        public string LeaseConnection { get; set; }
        public string LeaseDatabaseId { get; set; }
        public string LeaseContainerId { get; set; }
```

tomkerkhove commented on deploy/deploy-scaledobject.yaml:
```diff
@@ -0,0 +1,22 @@
+    name: <application-deployment-name>
+  triggers:
+    - type: external
+      metadata:
+        scalerAddress: keda-cosmosdb-scaler.keda:4050
+        connection: <connection-string-of-monitored-container-account>
+        databaseId: <database-id>
+        containerId: <container-id>
+        leaseConnection: <connection-string-of-lease-container-account>
+        leaseDatabaseId: <lease-database-id>
+        leaseContainerId: <lease-container-id>
+        processorName: <processor-name>
```
```suggestion
      metadata:
        scalerAddress: keda-cosmosdb-scaler.keda:4050
        connection: <connection-string-of-monitored-container-account>
        databaseId: <database-id>
        containerId: <container-id>
        lease:
          connection: <connection-string-of-lease-container-account>
          databaseId: <lease-database-id>
          containerId: <lease-container-id>
        processorName: <processor-name>
```

We could even do this already for database & container, but we can wait

JatinSanghvi replied:
I am not sure about this. For example, `lease` is not the right word here as we are not talking about a single lease; the lease container will be holding several leases, may be even for multiple monitored containers. What I am interested in is lease container, so the metadata representation might be `leaseContainer: { id: "", databaseId: "", connection: ""}`. The `processorName` is name of lease container change-feed processor, so it's associated with lease container but it cannot be part of lease container definition. That makes it difficult to choose whether to place it outside or inside the `lease` object. I tried taking some reference from [WebJobs extension](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/main/src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerAttribute.cs) where lease and monitored container properties are defined at same level. To me, the current placement of properties looks good enough.

JatinSanghvi replied:
@tomkerkhove - Do you think the following metadata would improve upon the existing one? If so, I can work on it in this or next week.

```
metadata:
  scalerAddress: keda-cosmosdb-scaler.keda:4050
  container:
    containerId: <container-id>
    databaseId: <database-id>
    connection: <connection>
  leaseContainer:
    id: <lease-container-id>
    databaseId: <lease-database-id>
    connection: <lease-connection>
  processorName: <processor-name>
  ```

tomkerkhove replied:
I  think that would be lovely!

tomkerkhove reviewed (approved):
LGTM, a big improvement - Thanks a ton! End-users can now better understand how it works and if it's a fit for them.

Just made a remark on streamlining the SO metadata but rest looks fine 👌

tomkerkhove commented:
@pragnagopa @ealsur Please take a final look

pragnagopa commented on .github/workflows/pr-build.yml:
```diff
@@ -0,0 +1,53 @@
+      - main
+
+env:
+  VERSION_TAG: experimental
```
use `preview` instead

pragnagopa reviewed (approved):
LGTM !

ealsur commented on src/Scaler.Demo/OrderGenerator/Program.cs:
```diff
@@ -0,0 +1,147 @@
+        {
+            Console.WriteLine($"Creating database: {_cosmosDbConfig.DatabaseId}");
+
+            Database database = await new CosmosClient(_cosmosDbConfig.Connection)
```
nit: Worth using a single CosmosClient instance instead of creating one in the Setup, another in CreateOrdersAsync and another in Teardown

JatinSanghvi replied:
This isn't a problem, since in one run of the application, either `SetupAsync` or `CreateOrdersAsync` or `TeardownAsync` will execute (that too once only).

tomkerkhove commented:
Ha, I have missed the last commit @JatinSanghvi.

Would it be possible to change `ghcr.io/kedacore/keda-cosmosdb-scaler` back to `ghcr.io/kedacore/keda-external-scaler-azure-cosmos-db` please?

This is important for our scaler approach where people need to understand that it's an external scaler and follow the convention. THank you!

pragnagopa commented:
> This is important for our scaler approach where people need to understand that it's an external scaler and follow the convention. THank you!

@JatinSanghvi  - please use same name in mcr as well.

JatinSanghvi commented:
> @JatinSanghvi  - please use same name in mcr as well.

Sure @pragnagopa, I was going to use the same name to avoid confusion.
